fix: Interpreter error formats, better handling

Cobalt0s opened the pull request:
# Bug

There is a concern that error resposne format from any connector can change and current error response parsing has no good fallback.

# Reviewing
First commit has the main implementation show casing usage on Pipeliner. Consequent commits update the rest of connectors.

# Implementation

As of now any error response will go via `interpretJSONError` method. Every connector has it.
Format switch allows to select the best response format that knows how to convert itself into an error.

```go
schema, err := interpreter.NewFormatSwitch(
 []interpreter.FormatTemplate{
  {
    MustKeys: []string{"description"},
    Template: &DescriptiveResponseError{},
  }, {
    MustKeys: []string{"status", "code"},
    Template: &StatusResponseError{},
  }, {
    MustKeys: nil,
    Template: &MessageResponseError{},
  },
 }...,
).ParseJSON(responseBody)
```
This can be understood as follows: if response payload has `description` key then `DescriptiveResponseError` will be used for JSON parsing and it knows how to produce go error. If `desription` is missing then it will look for both `status` and `code` keys and will then select `StatusResponseError` format. And lastly it will default to `MessageResponseError`.

If at any step JSON marshalling fails it will fallback to template that uses the whole response.Body as a message string. 

Conclusion, every message is exposed and if lucky it will use good formating.

jlimatampersand commented:
@Cobalt0s overall looks great, but I think we can further optimize this to reduce number of lines and manual coding errors. 

We we define statusCodeMapping as map format and add those in initialization time in `NewConnector()` call. Same for format interpreter as below

```
func NewConnector() {
  conn := &SalesloftConnector{}
  ...
  ... 
  //some init goes here
  ...
  conn.statusCodeMap = map[http.StatusCode] error
  conn.errorTemplates = []Templates{someTemplates...,...} 
  
}
```

Also we define statusCodeMapping only once for all connector as below
```
func statusCodeMapping[T Connector] (conn *T, res *http.Response, body []bytes) error {
    if conn.statusCodeMap == nil {
      return interpreter.DefaultStatusInterpreter(res, body)
    }

    mappedErr, ok := conn.statusCodeMap[res.StatusCode]
    if !ok {
      return interpreter.DefaultStatusInterpreter(res, body)
    }
     
    return mappedErr

}
```

The for all of the `interpretJSONError` function will look the same as below 

```
func (conn *SalesloftConnector) interpretJSONError (res, body) error {
  schema,err := conn.GetErrorFormatter().ParseJSON(body)
  if err != nil {
   return err
  }
  
  return schema.CombineErr(statusCodeMapping[SalesloftConnector](conn, res, body))

}

```

In fact we can also improve this by doing below
```
func (conn *SalesloftConnector) interpretJSONError (res, body) error {
  return InterpretJsonError[SalesloftConnector](conn, res, body)
}
```

Does this kind of approach make sense to you? This will reduce number of repetitive lines for all connectors and will reduce human code error. 

Let me know how you think about this approach.

Cobalt0s commented on common/interpreter/formatTemplate.go:
```diff
@@ -0,0 +1,31 @@
+	MustKeys []string
+	// Template is a struct pointer which will be used to flush the data into.
+	// Must implement ErrorDescriptor.
+	Template any
```
For some reason I had issues with interface unmarshalling. I did debugging and it seems that things have changed for the better. Anyway, I will be changing to explicit interface.

Cobalt0s commented:
@jlimatampersand this is great, I like your thoughts! I revisited the code and introduced further code optimization that you have suggested.

## New commits

First of all, this commit addresses your general comments: https://github.com/amp-labs/connectors/pull/840/commits/dd771e5f93466d5c1e0c0a695f0ef3a8862a4388

Then I altered some interpreter interfaces: https://github.com/amp-labs/connectors/pull/840/commits/c430f2819261be579294ac097c095df7ab1ab021
I have not used any generics. One problem is that interpreter package shouldn't know/manage data that lives on connector struct. Hence, I took close direction. Under this commit you will also see the usage on Pipeliner.

Next commits follow what Pipeliner did. I have removed some test files `errors_test.go` as I don't see significant value in them. This is when I introduced any testing and `read_test.go` and other do better job at covering the error handling. Maybe interpreter package could have more tests.

## Connector Init

The connector setup looks as follows:
```
conn.Client.HTTPClient.ErrorHandler = interpreter.ErrorHandler{
	JSON: interpreter.NewFaultyResponder(errorFormats, statusCodeMapping),
}.Handle
```
> Note: HTTPClient is coming from `WithClient()`. Maybe we could think how to take this connector parameter and configure with error handling, base url in more readable form. It is refactored code from way ago, and it seems we can do better.

Alternative connector initialization where error is handled in a non uniform way (salesforce):
```
conn.Client.HTTPClient.ErrorHandler = interpreter.ErrorHandler{
	JSON: &interpreter.DirectFaultyResponder{Callback: conn.interpretJSONError},
	XML:  &interpreter.DirectFaultyResponder{Callback: conn.interpretXMLError},
}.Handle
```

## Benefits
For people that will implement a connector it will be a matter of specifying:
* list of possible error formats (optional, but will yield better message formating control)
* unique status code mappings (if any)
This means only data driven, so configuration like.
By design, I kept the handler as interface so anybody can plug in their implementation, just in case.

jlimatampersand commented:
LGTM! thanks for working on this!
